presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java

```diff
@@ -22,17 +22,54 @@
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.IntPredicate;
 
+import static com.facebook.presto.sql.planner.PartitioningScheme.Replication.REPLICATE_NOTHING;
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 public class PartitioningScheme
 {
+    public enum Replication
+    {
+        REPLICATE_NOTHING(
+                false,
+                columnsCount -> true),
+
+        REPLICATE_NULLS(
+                true,
+                columnsCount -> columnsCount <= 1);
+
+        private final boolean replicatesNulls;
```

Comment:
I'm thinking this is a bit overkill, especially with the "replicatesAny" that gets added later. I'm not convinced generalizing these aspects of replication makes things easier or more readable (vs just checking against the enum value wherever it matters). I'll play around with the code a bit and see how it ends up.